dune-precice/dune-precice/couplinginterface.hh

```diff
@@ -0,0 +1,264 @@
+// -*- tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
+// vi: set et ts=4 sw=2 sts=2:
+
+#ifndef DUNE_PRECICE_HH
+#define DUNE_PRECICE_HH
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <dune/istl/io.hh>
+#include <dune/istl/bvector.hh>
+
+#include <dune/functions/functionspacebases/interpolate.hh>
+
+#include <precice/SolverInterface.hpp>
+
+namespace Dune::preCICE {
+
+  namespace Utilities {
+
+    template <int dim, typename VectorType>
+    static void copy_from_precice_to_dune(VectorType &dune_data,
+                                          const std::vector<double> &precice_data,
+                                          const BlockVector<FieldVector<bool, dim>> is_coupling_boundary)
+    {
+      int iteration_count = 0;
+      for(int i=0; i<dune_data.N(); i++) {
+        for(int j=0; j<dim; j++) {
+          if(is_coupling_boundary[i][j]) { dune_data[i][j] = precice_data[iteration_count]; ++iteration_count; }
+        }
+      }
+    }
+  
+    template <int dim, typename VectorType>
+    static void copy_from_dune_to_precice(const VectorType &dune_data,
+                                          std::vector<double> &precice_data,
+                                          const BlockVector<FieldVector<bool, dim>> is_coupling_boundary)
+    { 
+      int iteration_count = 0;
+      for(int i=0; i<dune_data.N(); i++) {
+        for(int j=0; j<dim; j++) {
+          if(is_coupling_boundary[i][j]) { precice_data[iteration_count] = dune_data[i][j]; ++iteration_count; }
+        }
+      }
+    }
+  } // namespace Utilities  
+
+  template <int dim, typename VectorType, typename ParameterClass>
+  class CouplingInterface
+  {
+
+    private:
+    
+      precice::SolverInterface precice;
+      
+      const BlockVector<FieldVector<bool, dim>> is_coupling_boundary_;
+
+      const std::string mesh_name_;
+      const std::string read_data_name_;
+      const std::string write_data_name_;
+    
+      int mesh_id_;
+      int read_data_id_;
+      int write_data_id_;
+      
+      int n_interface_nodes_;
+      std::vector<double> interface_nodes_positions_;
+      std::vector<int>    interface_nodes_ids_;
+      
+      std::vector<double> read_data_;
+      std::vector<double> write_data_;
+      
+      std::vector<VectorType> old_state_data_;
+      double old_time_;
+      int old_iter_;
+     
+      int mpi_rank_;  
+
+    public:
+              
+      CouplingInterface(const ParameterClass &parameters, 
+                        const BlockVector<FieldVector<bool, dim>> &is_coupling_boundary,
+                        int mpi_rank, int mpi_size);
+
+      template <typename PolynomialBasis>
+      double initialize(const PolynomialBasis &polynomial_basis);
+       
+      void read_blockvector_data(VectorType &dune_data);
+      void send_blockvector_data(const VectorType &dune_data);
```

Comment:
```suggestion
      void write_blockvector_data(const VectorType &dune_data);
```
Why is it called `send_blockvector_data`? It technically writes the data into the buffer. Data is only sent when `advance` is called. I think it would be possible to call the `write` functions of preCICE several times and even overwrite data in the buffer before data is actually send.

Reply:
Agreed, write is a way better naming for it!